### ISSUE TITLE
Dynamic Unions with GenericAliases failing

### ISSUE THREAD

**Issue by art049** [pydantic/pydantic#8362]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Hello,
I'm willing to define dynamic unions using the `typing.GenericAlias` .

The following:
```python
from typing import GenericAlias, Union
from pydantic import BaseModel, Field

class WithGenericAlias(BaseModel):
    a: GenericAlias(Union, (int, None)) = Field(3, gt=0)
```

Fails with this error: `TypeError: Plain typing.Union[int, None] is not valid as type argument` while this was handled with Pydantic v1

For a bit of context, I'm working on an ODM, and being able to patch the types(add annotations, for example) before building the underlying Pydantic model is necessary.

Strangely, removing the field constraint `gt=0` prevents the error from happening.

I already had some issues related to GenericAliases equality since `GenericAlias(Union, (int, None)) != Union[int, None]`, so this might be a part of the reason why this fail but I'm unsure.

### Workaround
If anyone faced the issue, I found a (super dirty) workaround that can work: 
```python
NewG = Union[int, str]  # We don't care about int,str since they will be replaced
setattr(NewG, "__args__", my_union_type_tuple)
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.5.2
        pydantic-core version: 2.14.5
          pydantic-core build: profile=release pgo=true
                 install path: /home/arthur/.pyenv/versions/3.10.4/envs/odmantic/lib/python3.10/site-packages/pydantic
               python version: 3.10.4 (main, Jul 22 2022, 14:35:40) [GCC 12.1.0]
                     platform: Linux-5.15.142-1-MANJARO-x86_64-with-glibc2.38
             related packages: typing_extensions-4.9.0 fastapi-0.104.1 mypy-1.4.1
```

**Comment by Viicos** [comment]
I believe this should work as expected, you are just hitting the warning [here](https://docs.pydantic.dev/latest/concepts/fields/#numeric-constraints) (see _Constraints on compound types_). Perhaps `a: GenericAlias(Union, (Annotated[int, Field(gt=0)], None)) = 3`?

**Comment by art049** [comment]
Ah, yes, totally, my bad. Is there an issue tracking this? Or is it not planned to be supported?

**Comment by Viicos** [comment]
I think it isn't planned, as it is better to be explicit about the constraints applied to types. The issue comes up pretty often though, so maybe a better error message could be implemented

**Comment by art049** [comment]
Yes, or maybe a warning when this undefined behavior can appear?

**Comment by sydney-runkle** [comment]
Going to mark this as a feature request then :). @art049, any interest in submitting a PR with support for a warning or better error message? I also think we could expand on that section of the docs a bit.

**Comment by art049** [comment]
Gonna try to find some time to implement it!

**Comment by sydney-runkle** [comment]
> I also think we could expand on that section of the docs a bit.

Could be nice to move that warning to a more central location in the docs (or perhaps, a common mistakes section).

**Comment by Viicos** [comment]
> Could be nice to move that warning to a more central location in the docs (or perhaps, a common mistakes section).

Yes, couldn't find a better place when adding this section
